mypy/mixedtraverser.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional
+
+from mypy.nodes import (
+    MypyFile, Var, FuncItem, ClassDef, AssignmentStmt, OperatorAssignmentStmt, ForStmt, WithStmt,
+    CastExpr, TypeApplication, TypeAliasExpr, TypeVarExpr, TypedDictExpr, NamedTupleExpr,
+    EnumCallExpr, PromoteExpr, NewTypeExpr
+)
+from mypy.types import Type
+from mypy.traverser import TraverserVisitor
+from mypy.typetraverser import TypeTraverserVisitor
+
+
+class MixedTraverserVisitor(TraverserVisitor, TypeTraverserVisitor):
+    """Recursive traversal of both Node and Type objects."""
+
+    # Symbol nodes
+
+    def visit_var(self, var: Var) -> None:
+        self.visit_optional_type(var.type)
+
+    def visit_func(self, o: FuncItem) -> None:
+        super().visit_func(o)
+        self.visit_optional_type(o.type)
+
+    def visit_class_def(self, o: ClassDef) -> None:
+        super().visit_class_def(o)
+        info = o.info
+        if info:
+            for base in info.bases:
+                base.accept(self)
+            self.visit_optional_type(info.typeddict_type)
+
+    def visit_type_alias_expr(self, o: TypeAliasExpr) -> None:
+        super().visit_type_alias_expr(o)
+        o.type.accept(self)
+
+    def visit_type_var_expr(self, o: TypeVarExpr) -> None:
+        super().visit_type_var_expr(o)
+        o.upper_bound.accept(self)
+        for value in o.values:
+            value.accept(self)
+
+    def visit_typeddict_expr(self, o: TypedDictExpr) -> None:
+        super().visit_typeddict_expr(o)
+        self.visit_optional_type(o.info.typeddict_type)
+
+    def visit_namedtuple_expr(self, o: NamedTupleExpr) -> None:
+        super().visit_namedtuple_expr(o)
+        assert o.info.tuple_type
+        o.info.tuple_type.accept(self)
+
+    def visit__promote_expr(self, o: PromoteExpr) -> None:
+        super().visit__promote_expr(o)
+        o.type.accept(self)
+
+    def visit_newtype_expr(self, o: NewTypeExpr) -> None:
+        super().visit_newtype_expr(o)
+        self.visit_optional_type(o.old_type)
```

Comment:
The type that appears in `bases` of the generated `TypeInfo` may be a copy of this one, we had a crash because of this a while ago.

Reply:
`bases` actually don't seem to be visited at all. This should be the only place where we visit the base type. More generally, generated TypeInfos are handled in a bit ad hoc fashion, but it shouldn't be a problem for type argument checks, as it's somewhat reasonable to assume that generated things are valid. I'll create a follow-up issue about this.